Fix arrow keys in C64 controller mapping & other libretro-vice settings

NostalgicDavid opened the pull request:
Remove datasette actions from C64 controller mappings.
This fixes arrow keys.

@Kugelblitz360 has thinks arrow keys should not be mapped to datasette actions.

sonninnos commented:
Those keys do nothing if `vice_datasette_hotkeys` is disabled, which it is by default in the core, but you have it enabled:

```
coreSettings.save('vice_datasette_hotkeys', '"enabled"')
```

There is zero reason to have that one enabled, and have those hotkeys nulled.

Kugelblitz360 commented:
That sounds like a valid explanation. So we should have the hotkeys disabled as per sonninnos suggestion and not by pulling them completely.

nadenislamarre commented:
so, should i merge or not merge ?

Kugelblitz360 commented:
Please do not merge this one. I will aim to consolidate all changes once the latest vice-libretro works in a new V33 build (currently broken) based on all the stuff that @sonninnos integrated there (like Auto-Zoom).

NostalgicDavid commented:
Should I make vice_datasette_hotkeys configurable Batocera advanced setting for libretro-vice system?

Kugelblitz360 commented:
David, let's discuss a complete overhaul/merge once BetaV33 and the current version of the core are up and running and we can test the best config. But I would not make the vice_datasette_hotkeys configurable in EmuStation. Turning them off is correct for 99.9% of Bato usecases and too many options in ES is not a good idea. Whoever does specialized real tape emulation as a user most likely will not use Batocera but other distributions like Combian or BMC64. Bato is for playing games fast and easy, not complete esoteric tape loader support.

sonninnos commented:
> Does vice_datasette_hotkeys and vice_mapper_* settings affect all libretro-vice emulators (x64, x128 and others)?

Sure, so if one wants totally separate options for each core, one should not have `global_core_options` enabled in RA.

Indeed Datasette is very usable via the virtual keyboard anyway, so those hotkeys aren't that important.

iconoclusterdotexe commented:
> David, let's discuss a complete overhaul/merge once BetaV33 and the current version of the core are up and running and we can test the best config. But I would not make the vice_datasette_hotkeys configurable in EmuStation. Turning them off is correct for 99.9% of Bato usecases and too many options in ES is not a good idea. Whoever does specialized real tape emulation as a user most likely will not use Batocera but other distributions like Combian or BMC64. Bato is for playing games fast and easy, not complete esoteric tape loader support.

I fully agree with you.

@NostalgicDavid leave those arrow keys remap and just remove the vice_datasette_hotkeys line. If someone want it, it can be enabled on RA vice core options. Use this same PR without open a new one.

Kugelblitz360 commented:
So looking at the X64 version from Dec 30th including the latest greates vice-libretro, I would make the following suggestions regarding default settings, key mappings and Emulation Station fixes. All listed below are changes from the current defaults.
•	**Disable JiffyDOS as Default**. Batocera can not include the JiffyDOS ROMs because they are still copyrighted and actively sold (http://www.jiffydos.com/) (and they are not included anyway). Turning the Option ON when ROMs are not there may have side effects.
•	**Automatic Load Warp should be enabled**. Currently we load slow which is strange - even if JiffyDOS would be included.
•	**Global Work Disk should be enabled**. Some games write High Scores to disk even when loaded from tape/cartridge or as a PRG and not having a Global Work Disk might lead to crashes.
•	**Datasette Hotkeys should be disabled**. (As discussed at length).
•	Zoom Mode has a new Option, not yet in Emulation Station, „Auto Disable“, and this should be the new Default (and it should also be visible in Emulation Station - or we remove the setting completely from Emulation Station). It zooms in UNLESS there are border graphics, then it automatically zooms out – dynamically in game! (Confirmed with „Hunters Moon“). This is the preferred mode for all conceivable situations and I personally would never change back unless there is an incompatible game. But @sonninnos really added a fantastic option here and it works awesome!

This new Zoom mode would make R3 (Right thumbstick) free again (it currently has Zoom mapped to it via mapping it to F12 which is mapped to Zoom currently). Any Suggestion for a Quality of Life function we should map to R3?

iconoclusterdotexe commented:
> So looking at the X64 version from Dec 30th including the latest greates vice-libretro, I would make the following suggestions regarding default settings, key mappings and Emulation Station fixes. All listed below are changes from the current defaults. • **Disable JiffyDOS as Default**. Batocera can not include the JiffyDOS ROMs because they are still copyrighted and actively sold (http://www.jiffydos.com/) (and they are not included anyway). Turning the Option ON when ROMs are not there may have side effects. • **Automatic Load Warp should be enabled**. Currently we load slow which is strange - even if JiffyDOS would be included. • **Global Work Disk should be enabled**. Some games write High Scores to disk even when loaded from tape/cartridge or as a PRG and not having a Global Work Disk might lead to crashes. • **Datasette Hotkeys should be disabled**. (As discussed at length). • Zoom Mode has a new Option, not yet in Emulation Station, „Auto Disable“, and this should be the new Default (and it should also be visible in Emulation Station - or we remove the setting completely from Emulation Station). It zooms in UNLESS there are border graphics, then it automatically zooms out – dynamically in game! (Confirmed with „Hunters Moon“). This is the preferred mode for all conceivable situations and I personally would never change back unless there is an incompatible game. But @sonninnos really added a fantastic option here and it works awesome!
> 
> This new Zoom mode would make R3 (Right thumbstick) free again (it currently has Zoom mapped to it via mapping it to F12 which is mapped to Zoom currently). Any Suggestion for a Quality of Life function we should map to R3?
> 
> Somebody mentioned the Virtual Keyboard is not accessible from the keyboard. While this sounds strange at first, actually because of the datasette keys it is quite useful to have access to the VKBD from a physical keyboard. As mentioned above, F12 is now free because we do not need Zoom on a key anymore.
> 
> Finally as a bonus track, we might want to add specific settings to the Giana Sisters ROM included in the base image to have a SPECIFIC remapping of "Up" to Button B. This was a system wide default in V32 and below which actually breaks some other games, but it is really useful for this game. Please do not re-enable system wide, Button B as Space is usable with quite a lot of games.

If the next libretro-vice supports Auto-Zoom like PUAE, we can add it as an "Auto" choice to Emulationstation (I did the same with PUAE and added a lot of stuff like CPU Compatibility, Speed, Clock, etc...)

sonninnos commented:
> • **Disable JiffyDOS as Default**. Batocera can not include the JiffyDOS ROMs because they are still copyrighted and actively sold (http://www.jiffydos.com/) (and they are not included anyway). Turning the Option ON when ROMs are not there may have side effects.

No harm should be caused by having it enabled without the files. But if there is something, I will fix it.


> • **Global Work Disk should be enabled**. Some games write High Scores to disk even when loaded from tape/cartridge or as a PRG and not having a Global Work Disk might lead to crashes.

Don't! I haven't found a single game that would or could save scores to unit 9, and work disk does nothing in unit 8 if the core is launched with content. Only when starting the core without content will work disk in unit 8 get inserted. But nothing bad should happen anymore even if there is a work disk in unit 9 always, and nothing good either.

PRGs that can save scores should not be separated away from a disk image.


> Finally as a bonus track, we might want to add specific settings to the Giana Sisters ROM included in the base image to have a SPECIFIC remapping of "Up" to Button B. This was a system wide default in V32 and below which actually breaks some other games, but it is really useful for this game. Please do not re-enable system wide, Button B as Space is usable with quite a lot of games.

Simply set `vice_retropad_options` to `rotate_jump`, as in "Y = Fire, B = Up". That option acts as the "base", so any overlapping remapping option will override the buttons.

Kugelblitz360 commented:
As far as I can remember, VICE inserts the Workdisk when you directly inject PRGs or autostart CRTs? Yes, you are right that they should not be seperated from the disc, but there is stuff in the wild...

sonninnos commented:
If the work disk option is set to unit 8, it will never be inserted no matter what the format of the launched content is, if there is content. Technically it of course could do it when launching anything other than disks, since the unit will be free, but not currently, and the need is so tiny to keep motivation up.. And if the unit is 9, pretty much nothing (without manual intervention) will even see that it is there.

NostalgicDavid commented:
@Kugelblitz360 @sonninnos

- Automatic Load Warp is enabled.
- Datasette Hotkeys is disabled and mappings removed. AFAIK these mappings are defaults when datasette hotkeys are enabled. I don't see any reasons for making defaults hardcoded in this circumstances.
- Zoom Mode auto_disable is default. It is "Auto" and "Automatic-Disable" in ES. vice_zoom_mode=auto is "Automatic" in ES. vice_zoom_mode=manual cannot be set in ES. Zoom mode is unmapped from R3 and F12.
- F12 is mapped to Toggle Virtual Keyboard. R3 is unmapped from F12.
- JiffyDOS as Default is unchanged. It is still enabled.
- Global Work Disk is unchanged. vice_work_disk is not set.

How to make specific settings to the Giana Sisters ROM?

iconoclusterdotexe commented:
> @Kugelblitz360 @sonninnos
> 
> - Automatic Load Warp is enabled.
> - Datasette Hotkeys is disabled and mappings removed. AFAIK these mappings are defaults when datasette hotkeys are enabled. I don't see any reasons for making defaults hardcoded in this circumstances.
> - Zoom Mode auto_disable is default. It is "Auto" and "Automatic-Disable" in ES. vice_zoom_mode=auto is "Automatic" in ES. vice_zoom_mode=manual cannot be set in ES. Zoom mode is unmapped from R3 and F12.
> - F12 is mapped to Toggle Virtual Keyboard. R3 is unmapped from F12.
> - JiffyDOS as Default is unchanged. It is still enabled.
> - Global Work Disk is unchanged. vice_work_disk is not set.
> 
> How to make specific settings to the Giana Sisters ROM?

Don't add "auto" value on es_features at the moment. It needs a small fix to be working or it goes to conflict with ES "auto".

iconoclusterdotexe commented:
This fixes the issue:

libretroOptions.py:

```
        # Zoom Mode
        if system.isOptSet('zoom_mode_c64'):
            coreSettings.save('vice_zoom_mode', system.config['zoom_mode_c64'])
            if system.config['zoom_mode_c64'] == 'automatic':
                coreSettings.save('vice_zoom_mode', '"auto"')
        else:
            coreSettings.save('vice_zoom_mode', '"auto_disable"')
            
```
            
Now the "auto" value on core has renamed to "automatic"

es_features.yml;

```
            zoom_mode_c64:
                prompt:      ZOOM (HIDE BORDERS)
                description: Hides borders on many games. Some games used the borders.
                choices:
                    "Automatic-Disable":        auto_disable
                    "Automatic":                automatic
                    "Off":                      disabled
                    "small":                    small
                    "medium":                   medium
                    "maximum":                  maximum
                    
```

Kugelblitz360 commented:
Okay, regarding Giana Sisters, there are two possible solutions:

a) Expose the controller settings in Emulation Station as a menu item (vice_retropad_options). Then we can save it in the metadata, i.e. a modified gamelist.xml. Then it can also easily applied to other Jump-oriented games (e.g. Tower Toppler/Nebulus)
b) Give it a pad2key setting (are these also stored per game in gamelist.xml?) but then it is a special hack just for this one game.

Also we could exchange Giana Sisters with a new modified version that fixes NTSC? https://csdb.dk/release/?id=212344

BUT: Giana Sisters is copyrighted software and I am unsure whether the inclusion of it in Batocera is a good idea anyway. Same for the Super Mario conversion. Nintendo does not like when people use Mario. Would hate if this becomes an angle of attack.

iconoclusterdotexe commented:
> Okay, regarding Giana Sisters, there are two possible solutions:
> 
> a) Expose the controller settings in Emulation Station as a menu item (vice_retropad_options). Then we can save it in the metadata, i.e. a modified gamelist.xml. Then it can also easily applied to other Jump-oriented games (e.g. Tower Toppler/Nebulus)
> b) Give it a pad2key setting (are these also stored per game in gamelist.xml?) but then it is a special hack just for this one game.
> 
> Also we could exchange Giana Sisters with a new modified version that fixes NTSC? https://csdb.dk/release/?id=212344
> 
> BUT: Giana Sisters is copyrighted software and I am unsure whether the inclusion of it in Batocera is a good idea anyway. Same for the Super Mario conversion. Nintendo does not like when people use Mario. Would hate if this becomes an angle of attack.

Why not add free games instead of copyrighted ones? They're just for demonstration usage and a clean distribution. Add games like Giana Sisters or Super Mario Bros are a big risk (especially SMB).
Search for Psytronik Software. They have made some free games.

NostalgicDavid commented:
@nadenislamarre Can you please check this info from @Kugelblitz360 about Giana Sisters? I think it is important! 
> BUT: Giana Sisters is copyrighted software and I am unsure whether the inclusion of it in Batocera is a good idea anyway. Same for the Super Mario conversion. Nintendo does not like when people use Mario. Would hate if this becomes an angle of attack.

@Kugelblitz360 What do you think about changes in this Pull Request? Do they look OK to you? 
Do other games need vice_retropad_options setting exposed in Emulation Station? If they need, I will add it!

@iconoclusterdotexe Kudos! I have fixed the "Automatic" setting. And renamed it and "Automatic-Disable" to "Auto zoom" and "Auto-disable zoom".

iconoclusterdotexe commented:
> @nadenislamarre Can you please check this info from @Kugelblitz360 about Giana Sisters? I think it is important!
> 
> > BUT: Giana Sisters is copyrighted software and I am unsure whether the inclusion of it in Batocera is a good idea anyway. Same for the Super Mario conversion. Nintendo does not like when people use Mario. Would hate if this becomes an angle of attack.
> 
> @Kugelblitz360 What do you think about changes in this Pull Request? Do they look OK to you? Do other games need vice_retropad_options setting exposed in Emulation Station? If they need, I will add it!
> 
> @iconoclusterdotexe Kudos! I have fixed the "Automatic" setting. And renamed it and "Automatic-Disable" to "Auto zoom" and "Auto-disable zoom".

So, by adding the same function as in PUAE (called in "Jump on B" or "Jump on A" in the case of the CD32. Called this way to make it easier to understand with your current controller mapping). In fact, the values ​​for only "B = Fire, A = 2nd fire" and "B = Fire, A = Up" have been added without adding the values ​​of the Y button, as it has already been mapped to another function and could just create controller issues.

iconoclusterdotexe commented:
But the problem is: if we can add this, what button we can map as SPACE?

NostalgicDavid commented:
@Kugelblitz360 I think you can answer @iconoclusterdotexe question better than me.

Kugelblitz360 commented:
I had N on X for several games with "trainers" in front of them that ask you many y/n questions, but that was quality of life only, nothing essentials. Putting Space on X is a good idea. Space is often used as a secondary button in games and people have put the C64 on the floor so they can press the Space Bar with their feet in order to keep both hands on the joystick. Putting Space on X really makes sense to keep A/B selectable with an emulator option rather than keypad mapping.
For Y I would suggest either  the Statusbar toggle or Turbo Fire.

iconoclusterdotexe commented:
> I had N on X for several games with "trainers" in front of them that ask you many y/n questions, but that was quality of life only, nothing essentials. Putting Space on X is a good idea. Space is often used as a secondary button in games and people have put the C64 on the floor so they can press the Space Bar with their feet in order to keep both hands on the joystick. Putting Space on X really makes sense to keep A/B selectable with an emulator option rather than keypad mapping. We would need to have the option for that visible in Emulation Station then please!
> For Y I would suggest either  the Statusbar toggle or Turbo Fire (but I am unsure whether Turbo Fire on/off is managed by a hotkey by the current implentation of the core).

You can open Statusbar by hold SELECT for one second and more of a second you can fast forward your game.
Maybe using Y for Turbo isn't a bad idea.

Kugelblitz360 commented:
Yes. I totally forgot the multiple use of Select. Y for Turbo Fire (not Warp) would be nice then.

iconoclusterdotexe commented:
Unfortunately there isn't any direct Toggle Turbo Fire to map on Gamepad. We need to map a unused key first, after that mapping the key to a button.

The question is what key we can map? F9-F12 functions are mapped like this:
F9 - Toggle statusbar
F10 - Joyport switch
F11 -  Warp mode
F12 - Virtual keyboard.

They're all necessary for a keyboard only playing. Any suggestions?

NostalgicDavid commented:
@Kugelblitz360 @iconoclusterdotexe Can we merge this now? And make new PR after discussions done?

iconoclusterdotexe commented:
> @Kugelblitz360 @iconoclusterdotexe Can we merge this now? And make new PR after discussions done?

Have you read last comments before doing this?

Add `X button = Space`, `Right CTRL = Toggle Turbo Fire` and `Y button = Right CTRL`, but test them with a pair of games.

Kugelblitz360 commented:
OK**, I tested this on the latest Beta** and these would be the final, consolidated changes we should do for optimization. I think @iconoclusterdotexe did most of these already, but there is ONE final change - changing Y and X _(because I did get them wrong actually, my fault, sorry, old Xbox guy. Space should be on Y(West) and Turbo Fire on X(North) for better accessibility. At least thats what my thumb says after trying with three games that use Space as 2nd fire)._
a) Expose "Retropad Face Button Option" to Emulation Station as shortened text "Button Options". Only offer the first two options "B=Fire, A=2nd Fire" and "B=Fire A=Up (Jump)". Make the 2nd option the default so the "bundled" games work well.
b) Set Y (WEST) to Space (not X (North))
c) Set Toggle Turbo Fire to Right CTRL, Set Turbo Fire to OFF
d) Set X (NORTH) to Right CTRL
e) Set A (EAST) and B (SOUTH) to unmapped (IMPORTANT)
f) Set "Datasette Hot Keys" to disabled
g) Set "Automatic Load Warp" to enabled
As this now consolidates everything discussed in this thread and is tested, we should merge after these final changes. Thank you all for your patience! V33 will rock for usable C64 emulation!

Kugelblitz360 commented:
(And then I need to touch SDL Vice configs to have identical mappings)

NostalgicDavid commented:
@Kugelblitz360 Done! Can you please review and merge?

iconoclusterdotexe commented:
@NostalgicDavid for the jump button, use the same description as used for Amiga PUAE or users can confuse it.

NostalgicDavid commented:
@Kugelblitz360 Should we change Emulation Station "Button Options" to "JUMP ON A" with settings "Off" and "On"? Default would be "On" and description would be "Makes second fire button press up instead."?

Kugelblitz360 commented:
I am fine with all that. _Please note that I have not made any code changes and can not merge._ But @nadenislamarre should merge once the text change @iconoclusterdotexe suggested and @NostalgicDavid confirmed has been made to esfeatures.yml

NostalgicDavid commented:
@Kugelblitz360 @iconoclusterdotexe Done!
@nadenislamarre Can you please merge?

iconoclusterdotexe commented:
I'll do the same with Amiga: Using Y as SPACE instead of X, to make it more user friendly.
